migrations/entitlements/migration.go

```diff
@@ -0,0 +1,90 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package entitlements
+
+import (
+	"github.com/onflow/cadence/migrations"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/sema"
+)
+
+type EntitlementsMigration struct {
+	Interpreter *interpreter.Interpreter
+}
+
+var _ migrations.Migration = EntitlementsMigration{}
+
+func NewEntitlementsMigration(inter *interpreter.Interpreter) EntitlementsMigration {
+	return EntitlementsMigration{Interpreter: inter}
+}
+
+func (EntitlementsMigration) Name() string {
+	return "EntitlementsMigration"
+}
+
+// Converts its input to an entitled type according to the following rules:
+// * `ConvertToEntitledType(&T) ---> auth(Entitlements(T)) &T`
+// * `ConvertToEntitledType(Capability<T>) ---> Capability<ConvertToEntitledType(T)>`
+// * `ConvertToEntitledType(T?) ---> ConvertToEntitledType(T)?
+// * `ConvertToEntitledType(T) ---> T`
+// where Entitlements(I) is defined as the result of T.SupportedEntitlements()
+func ConvertToEntitledType(t sema.Type) sema.Type {
```

Comment:
This defines the main logic for converting a "pre-entitlements" reference type into a "post-entitlements" one